In [15]:
import pandas as pd

# 1. Import des données

In [16]:
df_irve = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/eb76d20a-8501-400e-b336-d85724de5435'
)

print(f"Shape : {df_irve.shape}")
df_irve.sample(5)

/tmp/ipykernel_70637/1170431219.py:1: DtypeWarning: Columns (12,18,19,20,21,22,24,25,29,33) have mixed types. Specify dtype option on import or set low_memory=False.
  df_irve = pd.read_csv(


Shape : (213699, 52)


,nom_amenageur,siren_amenageur,contact_amenageur,nom_operateur,contact_operateur,telephone_operateur,nom_enseigne,id_station_itinerance,id_station_local,nom_station,...,datagouv_resource_id,datagouv_organization_or_owner,created_at,consolidated_longitude,consolidated_latitude,consolidated_code_postal,consolidated_commune,consolidated_is_lon_lat_correct,consolidated_is_code_insee_verified,consolidated_is_code_insee_modified
199728,ENGIE Mobilités Electriques,909073363.0,cyrille.charton@engie.com,Greenflux,support@greenflux.com,+31 615 11 3734,ENGIE Vianeo,FRVIAP101101,NaN,ENGIE Vianeo - A6 Achères Est,...,ac32a7a4-8a1f-473d-b607-803270cf67c8,engie-mobilites-electriques,2026-04-08T11:43:38.122000+00:00,2.574533,48.361836,77760.0,Achères-la-Forêt,True,True,True
143998,QPARK,NaN,NaN,IZIVIA,sav@izivia.com,+33(0)972668001,QPARK,FRQPKPQPRK50,FRSODPQPRK502,QPARK - PERPIGNAN - RÉPUBLIQUE,...,a2a9e3fb-2f0a-486f-b535-b7e3b912c2c1,izivia,2023-09-19T14:48:09.441000+00:00,2.895326,42.698224,NaN,Perpignan,True,True,False
1612,ELECTRIC 55 CHARGING,832489801.0,contact@e55c.com,ELECTRIC 55 CHARGING,contact@e55c.com,33975891501,ELECTRIC 55 CHARGING,FR55CP05220MLBP0MM1ERS,FR*55C*P05220*MLB*P0MM1ERS,PARKING DES POMMIERS - MONETIER-LES-BAINS,...,7a5acb37-32c9-48bf-a14f-cac23bb9aff0,plus-de-bornes,2023-08-21T10:10:50.570000+00:00,6.504843,44.977618,5220.0,Le Monêtier-les-Bains,True,True,False
132173,GROUPE INDIGO,NaN,NaN,GROUPE INDIGO,support.emobility.fr@group-indigo.com,NaN,Réseau de recharge du groupe Indigo,FRPKGP11262001,NaN,Halles,...,d445f73a-797f-46ea-98cc-2927ec7a018d,indigo-group,2025-08-13T13:39:08.570000+00:00,3.005447,43.181774,NaN,NaN,False,False,False
210923,YES55,NaN,NaN,YES55 | FR*Y55,jimmy.guillemin@yes55.fr,NaN,YES55,FRY55P5944558972455143070,1547698,YES55/1958000579915292674,...,61387a4e-22f7-4662-b241-d5cac4dd91fd,gireve-2,2023-03-24T14:32:54.036000+00:00,-0.346221,47.841511,NaN,NaN,False,False,False


Ce premier jeu de données est la base nationale des IRVE (Infrastructures de Recharge pour Véhicules Électriques).
Cette table comporte 212234 lignes pour 52 variables.

In [17]:
df_ve = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/90e0d717-deda-4bdc-9987-f82faac5bc93',
 encoding='latin-1'
)

print(f"Shape : {df_ve.shape}")
df_ve.sample(5)

Shape : (703545, 8)


/tmp/ipykernel_70637/2115914144.py:1: DtypeWarning: Columns (0,2) have mixed types. Specify dtype option on import or set low_memory=False.
  df_ve = pd.read_csv(


,CODGEO,LIBGEO,EPCI,LIBEPCI,DATE_ARRETE,NB_VP_RECHARGEABLES_EL,NB_VP_RECHARGEABLES_GAZ,NB_VP
183637,63245,MOZAC,200070753.0,CA Riom Limagne et Volcans,2022-06-30,68,0,4050
291470,62872,WAMIN,200044030.0,CC des Sept Vallées,2021-06-30,0,0,265
637338,57109,BREISTROFF-LA-GRANDE,245700695.0,CC de Cattenom et Environs,2024-12-31,42,0,905
596241,32065,LE BROUILH-MONBERT,200072320.0,CC Val de Gers,2022-03-31,1,0,302
447032,14119,CAGNY,200065589,CC Val ès Dunes,2022-12-31,26,0,1816


Ce second jeu de données contient des informations sur les voitures particulières immatriculées par commune et par type de recharge. Il contient l'historique des immatriculations pour chaque commune à différentes dates.
Cette table comporte 703545 lignes pour 8 variables. 

In [18]:
df_revenus = pd.read_csv(
 'https://www.data.gouv.fr/api/1/datasets/r/516130bc-4dcb-47f5-8347-ae96553c43ab',sep=';'
)

print(f"Shape : {df_revenus.shape}")
df_revenus.sample(5)

Shape : (34926, 57)


/tmp/ipykernel_70637/3484954150.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df_revenus = pd.read_csv(


,Nom géographique GMS,Code géographique,Libellé géographique,[DISP] Nbre de ménages fiscaux,[DISP] Nbre de personnes dans les ménages fiscaux,[DISP] Nbre d'unités de consommation dans les ménages fiscaux,[DISP] 1ᵉʳ quartile (€),[DISP] Médiane (€),[DISP] 3ᵉ quartile (€),[DISP] Écart interquartile (€),...,[DEC] 9ᵉ décile (€),[DEC] Rapport interdécile D9/D1,[DEC] S80/S20,[DEC] Iice de Gini,[DEC] Part des revenus d’activité (%),[DEC] dont part des salaires et traitements (%),[DEC] dont part des iemnités de chômage (%),[DEC] dont part des revenus des activités non salariées (%),"[DEC] Part des pensions, retraites et rentes (%)",[DEC] Part des autres revenus (%)
30238,Jouy-sur-Morin,77240,Jouy-sur-Morin,882.0,2128.0,1431.8,17670.0,23150.0,28630.0,10960.0,...,37080.0,"3,6",4.7,0.277,67.2,58.8,3.0,5.4,28.3,4.5
19144,Suzannecourt,52484,Suzannecourt,146.0,355.0,240.2,NaN,21800.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10029,Digny,28130,Digny,380.0,944.0,626.0,NaN,23270.0,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
12147,Noulens,32299,Noulens,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18896,Daillancourt,52160,Daillancourt,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Ce dernier jeu de données contient des informations sur les voitures particulières immatriculées par commune et par type de recharge.
Cette table comporte 34926 lignes pour 57 variables.

# 2. Jointure des tables

## 2.1 Choix de la variable de jointure

Les 3 jeu de données contiennent une variable indiquant le code géographique mais sous 3 noms différents :
- `code_insee_commune` pour df_irve
- `CODGEO` pour df_ve
- `Code géographique` pour df_revenus

La jointure sera faite sur ces variables.

## 2.2 Préparation de la variable de jointure

Le diagnostic met en évidence plusieurs points d’attention :
- Présence de valeurs manquantes dans `code_insee_commune` (df_irve)  
- Présence de longueurs de codes incohérentes :
  - `code_insee_commune` (df_irve) contient des codes de longueur 3, 6 et 7
  - `CODGEO` (df_ve) contient des codes de longueur 4
- Les variables de jointure ne sont pas uniques :
  - `code_insee_commune` (df_irve)
  - `CODGEO` (df_ve)

Les clés de jointure présentent des problèmes de qualité et de format.  
Il est nécessaire de les nettoyer et les uniformiser (type, format, gestion des valeurs manquantes, unicité) avant d’effectuer la jointure.

### 2.2.1 Unification des codes

In [19]:
from src.preparation_data import nettoyer_code_insee

In [20]:
df_irve["code_insee_commune"] = df_irve["code_insee_commune"].apply(nettoyer_code_insee)
df_ve["CODGEO"] = df_ve["CODGEO"].apply(nettoyer_code_insee)
df_revenus["Code géographique"] = df_revenus["Code géographique"].apply(nettoyer_code_insee)

Les codes insee sont maintenant au même format dans les 3 bases de données : un code à 5 caractères.

### 2.2.2 Compléter les valeurs manquantes

In [21]:
from src.preparation_data import creer_gdf_irve, charger_communes, joindre_communes, ajouter_codes_geo

In [22]:
gdf_irve = creer_gdf_irve(df_irve, "consolidated_longitude", "consolidated_latitude")
gdf_communes = charger_communes()
gdf_result = joindre_communes(gdf_irve, gdf_communes)
df_irve = ajouter_codes_geo(df_irve, gdf_result)

There was an error while reading the file from the URL: https://minio.lab.sspcloud.fr/projet-cartiflette/production/provider=IGN/dataset_family=ADMINEXPRESS/source=EXPRESS-COG-CARTO-TERRITOIRE/year=2022/administrative_level=COMMUNE/crs=4326/DEPARTEMENT=20/vectorfile_format=geojson/territory=metropole/simplification=0/raw.geojson
Error message: '/vsimem/pyogrio_d1f29ac2ed2c421fa8ea786dd041a483' not recognized as being in a supported file format.; It might help to specify the correct driver explicitly by prefixing the file path with '<DRIVER>:', e.g. 'CSV:path'.


Nous avons maintenant 2 colonnes concernant le code géographique dans la base de données IRVE :
- `code_insee_commune` : la variable initiale
- `code_geo_total` : la variable recalculant tous les codes géographiques à partir de la latitude et la longitude

In [23]:
from src.preparation_data import compter_valeurs_manquantes, compter_uniques

In [24]:
colonnes_irve = ["code_insee_commune", "code_geo_total"]

print("Valeurs manquantes :")
print(compter_valeurs_manquantes(df_irve, colonnes_irve))

print("\nValeurs uniques :")
print(compter_uniques(df_irve, colonnes_irve))

Valeurs manquantes :
{'code_insee_commune': np.int64(64966), 'code_geo_total': np.int64(1818)}

Valeurs uniques :
{'code_insee_commune': 9440, 'code_geo_total': 10278}


Cette manipulation permet de passer de 57919 valeurs manquantes à 1768. De plus les codes recalculés sont plus fiables que ceux initiaux. En effet, les données entrées n'étant pas toujours vérifiées, des erreurs étaient présentes. Certaines code géographiques correspondaient au code postal par exemple.

## 2.3 Jointure

Réfléchissons aux variables à garder... Lesquelles pourraient être intéressante ?

In [25]:
list(df_irve.columns)

['nom_amenageur',
 'siren_amenageur',
 'contact_amenageur',
 'nom_operateur',
 'contact_operateur',
 'telephone_operateur',
 'nom_enseigne',
 'id_station_itinerance',
 'id_station_local',
 'nom_station',
 'implantation_station',
 'adresse_station',
 'code_insee_commune',
 'coordonneesXY',
 'nbre_pdc',
 'id_pdc_itinerance',
 'id_pdc_local',
 'puissance_nominale',
 'prise_type_ef',
 'prise_type_2',
 'prise_type_combo_ccs',
 'prise_type_chademo',
 'prise_type_autre',
 'gratuit',
 'paiement_acte',
 'paiement_cb',
 'paiement_autre',
 'tarification',
 'condition_acces',
 'reservation',
 'horaires',
 'accessibilite_pmr',
 'restriction_gabarit',
 'station_deux_roues',
 'raccordement',
 'num_pdl',
 'date_mise_en_service',
 'observations',
 'date_maj',
 'cable_t2_attache',
 'last_modified',
 'datagouv_dataset_id',
 'datagouv_resource_id',
 'datagouv_organization_or_owner',
 'created_at',
 'consolidated_longitude',
 'consolidated_latitude',
 'consolidated_code_postal',
 'consolidated_commune',
 '

In [26]:
list(df_ve.columns)

['CODGEO',
 'LIBGEO',
 'EPCI',
 'LIBEPCI',
 'DATE_ARRETE',
 'NB_VP_RECHARGEABLES_EL',
 'NB_VP_RECHARGEABLES_GAZ',
 'NB_VP']

In [27]:
list(df_revenus.columns)

['Nom géographique GMS',
 'Code géographique',
 'Libellé géographique',
 '[DISP] Nbre de ménages fiscaux',
 '[DISP] Nbre de personnes dans les ménages fiscaux',
 "[DISP] Nbre d'unités de consommation dans les ménages fiscaux",
 '[DISP] 1ᵉʳ quartile (€)',
 '[DISP] Médiane (€)',
 '[DISP] 3ᵉ quartile (€)',
 '[DISP] Écart interquartile (€)',
 '[DISP] 1ᵉʳ décile (€)',
 '[DISP] 2ᵉ décile (€)',
 '[DISP]3ᵉ décile (€)',
 '[DISP] 4ᵉ décile (€)',
 '[DISP] 6ᵉ décile (€)',
 '[DISP] 7ᵉ décile (€)',
 '[DISP] 8ᵉ décile (€)',
 '[DISP] 9ᵉ décile (€)',
 '[DISP] Rapport interdécile 9ᵉ décile/1ᵉʳ decile',
 '[DISP] S80/20',
 '[DISP] Iice de Gini',
 '[DISP] Part des revenus d’activité (%)',
 '[DISP] dont part des salaires et traitements(%)',
 '[DISP] dont part des iemnités de chômage (%)',
 '[DISP] dont part des revenus des activités non salariées (%)',
 '[DISP] Part des pensions, retraites et rentes (%)',
 '[DISP] Part des revenus du patrimoine et autres revenus (%)',
 "[DISP] Part de l'ensemble des prestat

### 2.3.1 Une ligne par code géo

Pour pouvoir efectuer la jointure sur les codes géographiques, il est essentiel que pour chacune de nos 3 bases de données, chaque ligne corresponde à un unique code géographique.
Il faut alors réfléchir à la façon dont on veut agréger df_irve et quelles variables on souhaite garder.
Il faut également étudier la base df_ve afin de supprimer les "doublons" de code géographique.

##### df_irve

Ici, l'objectif est de mesurer l'offre de recharge par commune. Comme il peut y avoir plusieurs points de recharge par commune, il faut agréger. L'objectif final est d'expliquer ou prédire le taux de véhicules électriques local.

Après étude des nos variables (cf etude_df_irve.ipynb) nous avons sélectionner plusieurs variables nous permettant d'en créer de nouvelles, agrégées par code géographique.

Certaines variables ont été écarté directement car jugées non pertinente pour notre sujet, d'autres présetaient trop de valeurs manquantes, d'autres étaient trop peu informatives ou encore était du texte de libre donc trop difficile à utiliser avec le peu de temps dont nous disposons.

variables utilisées (12 variables) pour la création des variables agrégées :`code_geo_total`,`nom_operateur`,
               'implantation_station',
               'nbre_pdc',
               'puissance_nominale',
               'prise_type_ef',
               'prise_type_2',
               'prise_type_combo_ccs',
               'prise_type_chademo',
               'gratuit',
               'paiement_cb',
               'paiement_autre'

variables créées : une ligne finale = un code géographique.

|variables utilisées| Variable finale   | Construction      |
|| ----------------- | ----------------- |
|| total_pdc         | somme             |
|| puissance_moyenne | moyenne           |
|| puissance_max     | max               |
|| nb_operateurs     | nunique           |
|| top_operateur     | mode              |
|| pct_type_2        | moyenne booléenne |
|| pct_gratuit       | moyenne booléenne |
|| part_voirie       | dummies + moyenne |

Compléter uniquement se tableau très proprement

Commençons par rendre les données des variables sélectionner propres avant de faire l'agrégation.

In [28]:
from src.nouvelles_variables import clean_irve_variables_finales

df_filtre = clean_irve_variables_finales(df_irve)
df_filtre.sample(5)

,code_geo_total,nom_operateur,implantation_station,nbre_pdc,puissance_nominale,prise_type_ef,prise_type_2,prise_type_combo_ccs,prise_type_chademo,gratuit,paiement_cb,paiement_autre,borne_rapide
211988,66171,Zephyre | FR*ZP1,Parking privé à usage public,1,7.4,False,True,False,False,False,False,True,False
98909,17306,GREENSPOT,Parking privé à usage public,10,22.0,True,True,False,False,<NA>,<NA>,<NA>,False
98421,90010,E.Leclerc | FR*LE2,Parking privé à usage public,10,22.0,True,True,False,False,False,False,True,False
166461,81287,Bouygues Energies & Services,Voirie,2,22.0,False,True,False,False,False,False,False,False
150428,66119,REVEO SYADEN | FR*RVE,Parking public,2,22.0,True,True,False,False,False,False,True,False


Nous pouvons maintenant procéder à la création des variables agrégées.

In [29]:
from src.nouvelles_variables import creer_features_irve

nv_var = creer_features_irve(df_filtre)
nv_var.sample(5)

,code_geo_total,total_pdc,puissance_moyenne,puissance_max,nb_operateurs,pct_type_2,pct_combo_ccs,pct_chademo,pct_type_ef,pct_gratuit,pct_paiement_cb,pct_paiement_autre,top_operateur,Parking privŽ rŽservŽ ˆ la clientle,Parking privŽ ˆ usage public,Parking privé réservé à la clientèle,Parking privé à usage public,Parking public,Station dédiée à la recharge rapide,Voirie
6775,63006,9,24.000000,25.0,2,0.333333,0.333333,0.333333,0.333333,0.0,1.0,1.0,SPIE CityNetworks,0.0,0.0,0.0,0.0,0.0,0.0,1.0
8337,77282,412,7.753846,25.2,1,0.923077,0.076923,0.0,0.769231,0.0,0.0,0.0,Bouygues Energies & Services,0.0,0.0,0.0,0.0,0.769231,0.0,0.230769
6975,64461,4,22.000000,22.0,1,1.0,0.0,0.0,1.0,0.0,1.0,0.0,inconnu,0.0,0.0,0.0,0.0,0.0,0.0,1.0
5161,50146,8,23.500000,25.0,1,0.5,0.5,0.0,0.0,0.0,1.0,0.0,STATIONS-E,0.0,0.0,0.0,0.0,1.0,0.0,0.0
9248,83099,506,64.352897,240.0,19,0.635514,0.35514,0.074766,0.252336,0.0,0.264706,0.979592,Allego,0.0,0.0,0.018692,0.336449,0.121495,0.233645,0.28972


In [38]:
nv_var.shape

(10278, 20)

In [39]:
list(nv_var.columns)

['code_geo_total',
 'total_pdc',
 'puissance_moyenne',
 'puissance_max',
 'nb_operateurs',
 'pct_type_2',
 'pct_combo_ccs',
 'pct_chademo',
 'pct_type_ef',
 'pct_gratuit',
 'pct_paiement_cb',
 'pct_paiement_autre',
 'top_operateur',
 'Parking priv\x8e r\x8eserv\x8e \x88 la client\x8fle',
 'Parking priv\x8e \x88 usage public',
 'Parking privé réservé à la clientèle',
 'Parking privé à usage public',
 'Parking public',
 'Station dédiée à la recharge rapide',
 'Voirie']

##### df_ve

La base véhicules contient plusieurs dates d'observation pour une même commune.

Afin d'être cohérent avec les autres sources de données (IRVE, INSEE), nous souhaitons disposer d'une seule ligne par commune correspondant à la situation la plus récente disponible.

Nous conservons donc, pour chaque code commune (`CODGEO`), la dernière observation temporelle.

In [30]:
from src.nouvelles_variables import preparer_base_ve

In [31]:
df_ve_latest = preparer_base_ve(df_ve)
print(df_ve_latest.shape)
df_ve_latest.head()

(35197, 5)


,CODGEO,DATE_ARRETE,NB_VP,NB_VP_RECHARGEABLES_EL,taux_equipement_ve
216883,01001,2025-09-30,968,30,0.030992
432383,01002,2025-09-30,286,5,0.017483
541297,01004,2025-09-30,16080,518,0.032214
109071,01005,2025-09-30,2306,68,0.029488
216903,01006,2025-09-30,169,4,0.023669


Deux variables sont conservées :

- `NB_VP` : nombre total de voitures particulières
- `NB_VP_RECHARGEABLES_EL` : nombre de voitures rechargeables / électriques

Plutôt que d'utiliser le nombre brut de véhicules électriques, nous construisons un ratio :

\[
taux\_equipement\_ve = \frac{NB\_VP\_RECHARGEABLES\_EL}{NB\_VP}
\]

Ce ratio est plus pertinent car il neutralise l'effet de taille des communes.

In [32]:
df_ve_latest["taux_equipement_ve"].describe().round(4)

count    35197.0000
mean         0.0229
std          0.0182
min          0.0000
25%          0.0128
50%          0.0205
75%          0.0302
max          1.0000
Name: taux_equipement_ve, dtype: float64

##### df_revenus

Voici notre première sélection de variables :

In [33]:
var_selec = ['Code géographique',
'[DISP] Médiane (€)',
 '[DISP] Iice de Gini',
 #garder uniquement l'un des deux
 '[DISP] Nbre de ménages fiscaux',
 '[DISP] Nbre de personnes dans les ménages fiscaux',
 #zone dynamique économiquement (potentiellement plus jeune aussi)
 '[DISP] Part des revenus d’activité (%)',
 #plus riche
 '[DISP] dont part des revenus des activités non salariées (%)',
 '[DISP] Part des revenus du patrimoine et autres revenus (%)'
]

df_revenus_final=df_revenus[var_selec]

**Variables à garder :**
[DISP] Médiane (€), [DISP] Part des revenus d’activité (%), [DISP] Iice de Gini.

Intérêt : La médiane du revenu est indispensable pour vérifier si l'adoption du VE est corrélée à la richesse de la commune. L'indice de Gini peut montrer si les communes très inégalitaires ont un comportement différent.

...

##### jointure

In [34]:
# jointure
df_final = (
    df_ve_latest
    .merge(nv_var, left_on="CODGEO", right_on="code_geo_total", how="left")
    .merge(df_revenus_final, left_on="CODGEO", right_on="Code géographique", how="left")
)

print("Shape base finale :", df_final.shape)

df_final.head()

Shape base finale : (35197, 33)


,CODGEO,DATE_ARRETE,NB_VP,NB_VP_RECHARGEABLES_EL,taux_equipement_ve,code_geo_total,total_pdc,puissance_moyenne,puissance_max,nb_operateurs,...,Station dédiée à la recharge rapide,Voirie,Code géographique,[DISP] Médiane (€),[DISP] Iice de Gini,[DISP] Nbre de ménages fiscaux,[DISP] Nbre de personnes dans les ménages fiscaux,[DISP] Part des revenus d’activité (%),[DISP] dont part des revenus des activités non salariées (%),[DISP] Part des revenus du patrimoine et autres revenus (%)
0,01001,2025-09-30,968,30,0.030992,NaN,NaN,NaN,NaN,NaN,...,<NA>,<NA>,01001,25820.0,NaN,346.0,895.0,NaN,NaN,NaN
1,01002,2025-09-30,286,5,0.017483,NaN,NaN,NaN,NaN,NaN,...,<NA>,<NA>,01002,24480.0,NaN,115.0,266.0,NaN,NaN,NaN
2,01004,2025-09-30,16080,518,0.032214,01004,542.0,129.367750,400.0,7.0,...,0.2,0.175,01004,21660.0,0.280,6855.0,15092.0,72.3,3.4,10.2
3,01005,2025-09-30,2306,68,0.029488,01005,8.0,26.666667,50.0,2.0,...,0.0,0.666667,01005,24610.0,0.246,800.0,2028.0,81.5,4.3,10.0
4,01006,2025-09-30,169,4,0.023669,NaN,NaN,NaN,NaN,NaN,...,<NA>,<NA>,01006,24210.0,NaN,51.0,107.0,NaN,NaN,NaN


In [37]:
# Nombre de valeurs manquantes par variable
df_final.isna().sum()

CODGEO                                                              0
DATE_ARRETE                                                         0
NB_VP                                                               0
NB_VP_RECHARGEABLES_EL                                              0
taux_equipement_ve                                                  0
code_geo_total                                                  24919
total_pdc                                                       24919
puissance_moyenne                                               24919
puissance_max                                                   24919
nb_operateurs                                                   24919
pct_type_2                                                      24919
pct_combo_ccs                                                   24919
pct_chademo                                                     24919
pct_type_ef                                                     24919
pct_gratuit         

Cette nouvelle base de données comporte 35197 communes.
Parmis ces communes 24919 ne contiennet pas de données provenant de df_irve.
Cela représente environ 70% de nos communes. Après étude de la réalité, nous considéront que ces 24919 sont sans point de charge. En effet ces chiffres sont crédible car une grande partie des communes françaises ont :
- peu d’habitants
- peu de commerces
- peu de parkings publics
- peu de flux routiers
donc l'absence de borne est très fréquente.

Les IRVE sont souvent concentrées dans :
villes
axes routiers
zones commerciales
parkings publics
intercommunalités centrales
Donc beaucoup de petites communes voisines n’en ont aucune.

Attention :
Une petite commune peut ne pas avoir de borne mais être à 5 km d’une commune équipée.
Donc absence locale ne signifie pas absence d’accès réel.

In [ ]:
df_imputation = df_final.copy()

# Liste des colonnes numériques de df_irve (comptage, puissance, pourcentages)
cols_to_zero = [
    'total_pdc', 
    'puissance_moyenne', 
    'puissance_max', 
    'nb_operateurs', 
    'pct_type_2', 
    'pct_combo_ccs', 
    'pct_chademo', 
    'pct_type_ef', 
    'pct_gratuit', 
    'pct_paiement_cb', 
    'pct_paiement_autre',
    'Parking priv\x8e r\x8eserv\x8e \x88 la client\x8fle', 
    'Parking priv\x8e \x88 usage public', 
    'Parking privé réservé à la clientèle', 
    'Parking privé à usage public', 
    'Parking public', 
    'Station dédiée à la recharge rapide', 
    'Voirie'
]

# 1. Remplacement par 0 pour les chiffres
df_imputation[cols_to_zero] = df_imputation[cols_to_zero].fillna(0)

# 2. Remplacement par un label pour le texte
df_imputation['top_operateur'] = df_imputation['top_operateur'].fillna('Aucun')

df_imputation.sample(5)

In [ ]:
missing = pd.DataFrame({
    "nb_manquants": df_imputation.isna().sum(),
    "pct_manquants": (df_imputation.isna().mean() * 100).round(2)
})

missing.sort_values("nb_manquants", ascending=False)

,nb_manquants,pct_manquants
[DISP] Iice de Gini,29854,84.82
[DISP] dont part des revenus des activités non salariées (%),29854,84.82
[DISP] Part des revenus du patrimoine et autres revenus (%),29854,84.82
[DISP] Part des revenus d’activité (%),29854,84.82
code_geo_total,24919,70.80
[DISP] Nbre de personnes dans les ménages fiscaux,3876,11.01
[DISP] Médiane (€),3876,11.01
[DISP] Nbre de ménages fiscaux,3876,11.01
Code géographique,275,0.78
CODGEO,0,0.00


[DISP] Iice de Gini
[DISP] dont part des revenus des activités non salariées (%)
[DISP] Part des revenus du patrimoine et autres revenus (%)
[DISP] Part des revenus d’activité (%)

Il va falloir enlever ces variables... trop de valeurs manquantes !

# 3. Analyse des variables

Après le choix des variables : 
0. variable cible = taux de VE
1. faire des analyses descriptives : 
- automatiser avec une fonction
- matrice de corrélation (heatmap)
2. modélisation :
simple regression ou plus avace (random forest...)

# 4. Modélisation

In [ ]:
from src.modelisation import modelisation_lasso, modelisation_xgboost

Nos variables pour la modélisation sont les suivantes :

attention LASSO ne prend pas de valeurs manquantes !!

In [ ]:
features = ['total_pdc', 'puissance_max', 'nb_operateurs', 'pct_gratuit']
target = 'taux_equipement_ve' 

df_model = df_final[features + [target]].dropna()

# 2. Exécuter le Lasso (Baseline & Sélection)
model_l, metrics_l, coeffs_l = modelisation_lasso(df_model, features, target)
print(f"R2 Lasso : {metrics_l['r2']:.3f}")
print("Déterminants (Lasso) :\n", coeffs_l[coeffs_l != 0])

# 3. Exécuter le Gradient Boosting (Performance)
model_x, metrics_x, importance_x = modelisation_xgboost(df_final, features, target)
print(f"R2 XGBoost : {metrics_x['r2']:.3f}")
importance_x.plot(kind='barh', title="Importance des variables - XGBoost")